refactor/runtime

char2cs opened the pull request:
## Overview
The `internal/infrastructure/runtime` package provides a **cross-platform process management system** for executing and managing OS-level processes. It follows Clean Architecture principles and provides a high-level abstraction over Go's `os/exec` package with enhanced features for process lifecycle management, output streaming, and concurrent process tracking.

```go
process, err := runtime.Get(ctx, "echo")
	.WithWorkDir("/tmp")
	.WithEnv(map[string]string{
		"FOO": "BAR",
	})
	.WithTimeout(10 * time.Second)
	.WithBufferSize(1024)
	.WithKillTimeout(10 * time.Second)
	.WithStopTimeout(10 * time.Second)
	.Build()

process.Start(ctx)
process.Stop(ctx)
process.Kill(ctx)
process.Wait(ctx)
```

## Architecture

### Core Components
**1. Runtime (Entry Point)**
- Main facade providing `Get()` and `Execute()` methods
- Handles OS detection and validation
- Delegates to the Manager for process tracking

**2. Manager**
- Central registry for all running processes
- Thread-safe process tracking using `sync.RWMutex`
- Bulk operations: `StopAll()`, `KillAll()`, `ShutdownAll()`
- Status-based filtering and cleanup of finished processes

**3. Process Interface & Implementations**
- **BaseProcess**: Common functionality shared across all OS implementations
- **DarwinProcess/LinuxProcess/WindowsProcess**: OS-specific signal handling and process group management
- Process states: `Prepared → Running → Stopping/Killing → Finished`

**4. Builder Pattern**
- Fluent API for process configuration: `WithEnv()`, `WithWorkDir()`, `WithBufferSize()`, etc.
- Automatic registration with Manager on build
- OS-specific process instantiation

**5. Output Handler**
- Dual-mode output capture: buffered (full output) + streaming (real-time)
- Separate stdout/stderr channels with configurable buffer sizes
- Thread-safe operations with mutex protection

**6. Models Package**
- `Config`: Process configuration (command, env, workdir, timeouts)
- `Status`: Process lifecycle state enumeration
- Custom error types for common failure scenarios

## Key Features
### Process Lifecycle Management
- Start, Stop (graceful with SIGTERM), Kill (forced with SIGKILL)
- Configurable timeouts for stop/kill operations
- Proper cleanup of pipes, channels, and OS resources

### Output Handling
- Real-time streaming via channels
- Full buffered output capture
- Separate stdout/stderr handling
- Scanner with 10MB line limit

### Concurrency Safety
- All operations protected by mutexes
- Safe concurrent access to Manager's process map
- WaitGroup synchronization for output goroutines

### Context Support
- Context cancellation for all blocking operations
- Timeout handling via context deadlines
- Graceful degradation on context cancellation

## Issues Identified & Fixed
### Critical Issues Fixed
**1. Race Condition in Output Handling**
- **Problem**: Output handler was closed before stdout/stderr goroutines finished draining pipes
- **Solution**: Added `sync.WaitGroup` to ensure goroutines complete before closing channels
- **Impact**: Eliminated data races in `TestManager_ShutdownAll` and related tests

**2. Test Timing Issue**
- **Problem**: `TestManager_ListByStatus` used fast-completing `echo` commands that transitioned to `Finished` before status checks
- **Solution**: Changed to long-running `sleep` commands with 100ms stabilization delay
- **Impact**: Test now reliably validates running process detection

**3. Environment-Specific Signal Handling**
- **Problem**: CI environments with sandboxing restrict process group signaling
- **Solution**: Added graceful skip for "operation not permitted" errors with reference to Manager-level tests
- **Impact**: Tests pass in both local and CI environments

## Test Coverage
- **Unit Tests**: All components (Manager, Process, Builder, Output, Models)
- **Integration Tests**: Cross-process interactions, concurrent operations
- **Race Detection**: All tests pass with `-race` flag
- **OS-Specific Tests**: Platform-specific skip logic for Darwin/Linux/Windows

## Design Patterns Used
1. **Facade Pattern**: Runtime provides simplified interface
2. **Builder Pattern**: Fluent process configuration
3. **Registry Pattern**: Manager tracks all processes
4. **Strategy Pattern**: OS-specific process implementations
5. **Observer Pattern**: Output streaming via channels

char2cs committed:
refactor(runtime): update runtime initialization and error handling

- Changed Runtime field in Infrastructure struct to a pointer.
- Enhanced error handling during runtime initialization to allow application functionality without runtime support.
- Removed unused Darwin and Linux runtime implementations and their associated tests to streamline the codebase.

char2cs commented:
@AgustinGil21 Esta es mi idea para este modulo, me vole un toque JAJAJAJAJ, pero decime que pensas

char2cs committed:
refactor(errors): remove unnecessary blank line in error definitions

char2cs committed:
refactor(tests): remove unnecessary blank lines in test files

AgustinGil21 commented:
@char2cs me parece muy buena las modificaciones que aplicaste pero decime si esta idea no te parece:

process, err := runtime.Get(ctx, "echo")
.Config(
  Dir: "/tmp",
  Env: map[string]string{
   "FOO": "BAR
  },
  ExecuteTimeout: 10 * time.Second,
  BufferSize: 1024
  KillTimeout: 10 * time.Second,
  StopTimeout: 10 * time.Second,
 ).Build()

process.Start(ctx)
process.Stop(ctx)
process.Kill(ctx)
process.Wait(ctx)

AgustinGil21 commented on internal/infrastructure/runtime/models/models.go:
```diff
@@ -0,0 +1,63 @@
+
+func (s Status) IsTerminal() bool {
+	return s == StatusFinished
+}
```
No se que te parece pero creo que IsFinished() sería mas descriptivo

AgustinGil21 commented on internal/infrastructure/runtime/output/output.go:
```diff
@@ -0,0 +1,116 @@
+	h.closeMu.Lock()
+	closed := h.closed
+	h.closeMu.Unlock()
+
+	if !closed {
+		select {
+		case h.errChan <- line:
+			// Successfully sent
+		default:
+			// Channel full, drop line to prevent blocking
+		}
+	}
```
Capaz aca un log estaría bueno, que sea un warn o debug por si se da un caso que haya un bug relacionado a eso

AgustinGil21 commented on internal/infrastructure/runtime/process/linux.go:
```diff
@@ -0,0 +1,121 @@
+		return models.ErrNoProcess
+	}
+
+	if err := p.cmd.Process.Kill(); err != nil {
+		if errors.Is(err, os.ErrProcessDone) {
+			select {
+			case <-p.doneChan:
+				return nil
+			case <-time.After(1 * time.Second):
+				return nil
+			}
+		}
+		return fmt.Errorf("failed to kill: %w", err)
+	}
```
El time.After no seria mejor que tenga un poco mas de tiempo? Capaz 1s es muy poco

char2cs committed:
refactor(runtime): rename IsTerminal to IsFinished and enhance output logging

- Updated the Status method from IsTerminal to IsFinished for clarity.
- Added logging for dropped lines in output channels to improve monitoring.
- Introduced a new method in Builder to set configuration, enhancing flexibility.
- Increased timeout duration for process termination across all platforms to ensure graceful shutdown.

char2cs committed:
refactor(tests): enhance netbridge and runtime tests for better coverage

- Added new test cases for port availability and forwarding in the netbridge implementation.
- Improved test structure for the runtime manager, including concurrent access and error handling scenarios.
- Updated existing tests to ensure they accurately reflect the current implementation and improve overall test coverage.
- Renamed test functions for clarity and consistency.

char2cs commented:
@AgustinGil21 ya estan los cambios que me pediste! + Ahora la CI no va a tirar error, el codigo de cada implementacion se va a compilar _solamente_ en la plataforma host. 85% de code coverage!
